projects/rx-form-mapper/src/lib/services/form-reader.ts

```diff
@@ -0,0 +1,78 @@
+import { InjectFlags, Injector, Type } from '@angular/core';
+import { AbstractControl, FormArray, FormControl, FormGroup } from '@angular/forms';
+import { CustomControlMapper } from '..';
+import { ControlVisitor, CustomControlMetadata, FormArrayMetadata, FormControlMetadata, FormGroupMetadata, FormMetadata } from '../metadata';
+
+export class FormReader implements ControlVisitor<any> {
+
+	public constructor(private readonly control: AbstractControl, private readonly injector: Injector) {}
+
+	public visitCustomControlMetadata(customControlMetadata: CustomControlMetadata): any {
+		return this.resolveCustomMapper(customControlMetadata.mapper).readForm(this.control);
+	}
+
+	public visitFormArrayMetadata(formArrayMetadata: FormArrayMetadata) {
+		if (!this.control) {
+			return void 0;
+		}
+
+		if (!(this.control instanceof FormArray)) {
+			throw new Error('controller is not FormArray instance');
```

Comment:
Non sarebbe meglio aver un enum Error? Così non metti stringhe all'interno della funzione

Reply:
Cioè? come l'avevi pensata?

Reply:
Ad esempio:

`enum FormReaderError { FormArrayInstanceError, FormGroupInstanceError, ... }`

Così da usare `FormReaderError.FormArrayInstanceError` nei vari throw.
In questo modo nel tuo codice sei parlante e più mantenibile in futuro, oltre al fatto che centralizzi i codici di errore.
Se ti torna meglio, potresti addirittura fare una classe che cattura gli errori per ciascuna feature così da avere anche errori dinamici, nel caso in cui in output devi mandare una variabile o altri log. In alternativa, enum è la via più semplice imho.

È una proposta, non c'è niente di troppo malvagio a lasciare una stringa

Reply:
fixato con un metodo a fattor comune :)